cirq/sim/state.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helpers for handling states."""
+
+import itertools
+
+from typing import Sequence, Union
+
+import numpy as np
+
+
+def pretty_state(state: Sequence, decimals: int=2) -> str:
+    """Returns the wavefunction as a string in Dirac notation.
+
+    For example:
+        state = np.array([1/np.sqrt(2), 1/np.sqrt(2)], dtype=np.complex64)
+        print(pretty_state(state)) -> 0.71|0⟩ + 0.71|1⟩
+
+    Args:
+        state: A sequence representing a wave function in which the ordering
+            mapping to qubits follows the standard Kronecker convention of
+            numpy.kron.
+        decimals: How many decimals to include in the pretty print.
+
+    Returns:
+        A pretty string consisting of a sum of computational basis kets
+        and non-zero floats of the specified accuracy.
+    """
+    perm_list = ["".join(seq) for seq in itertools.product(
+        "01", repeat=int(len(state)).bit_length() - 1)]
+    components = []
+    ket = "|{}⟩"
+    for x in range(len(perm_list)):
+        format_str = "({:." + str(decimals) + "g})"
+        # Python 2 rounds imaginary numbers to 0, so need to round separately.
+        val = (round(state[x].real, decimals)
+               + 1j * round(state[x].imag, decimals))
+        if round(val.real, decimals) == 0 and round(val.imag, decimals) != 0:
+            val = val.imag
+            format_str = "({:." + str(decimals) + "g}j)"
+        if round(val.imag, decimals) == 0 and round(val.real, decimals) != 0:
+            val = val.real
+        if val != 0:
+            if round(state[x], decimals) == 1:
+                components.append(ket.format(perm_list[x]))
+            else:
+                components.append((format_str + ket).format(val, perm_list[x]))
+
+    return ' + '.join(components)
+
+
+def decode_initial_state(initial_state: Union[int, np.ndarray],
+    num_qubits: int, dtype: np.dtype = np.complex64) -> np.ndarray:
```

Comment:
These arguments interplay in complicated ways. Should we infer `num_qubits` when `initial_state` is a vector?